File: templates/python/moving_average.py
"""moving_average: 原始 + 多窗口移动平均."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def make_figure(t=None, y=None, windows=(5, 20, 60), title='Moving averages'):
    apply_theme(fig_size=(8, 4))
    if t is None:
        rng = np.random.default_rng(2)
        t = np.arange(500); y = np.cumsum(rng.standard_normal(500)) + 0.01*t
    fig, ax = plt.subplots()
    ax.plot(t, y, color='lightgray', linewidth=0.8, label='raw')
    for i, w in enumerate(windows):
        ma = np.convolve(y, np.ones(w)/w, mode='same')
        ax.plot(t, ma, color=cycle(i), linewidth=1.5, label=f'MA({w})')
    ax.set_xlabel('t'); ax.set_ylabel('value'); ax.set_title(title)
    ax.legend(); ax.grid(True, linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure()
    fig.savefig(__file__.replace('.py', '.png'), dpi=150)

File: templates/matlab/moving_average.m
function fig = moving_average()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(2);
    t = 0:499; y = cumsum(randn(1, 500)) + 0.01*t;
    fig = figure('Position',[100 100 800 400]); hold on;
    plot(t, y, 'Color', [0.75 0.75 0.75], 'LineWidth', 0.8);
    for i = 1:3
        w = [5 20 60]; w = w(i);
        ma = filter(ones(1,w)/w, 1, y);
        plot(t, ma, 'Color', palette('cat',i), 'LineWidth', 1.5);
    end
    xlabel('t'); ylabel('value'); title('Moving averages');
    legend({'raw','MA(5)','MA(20)','MA(60)'}); grid on;
end
